### ISSUE TITLE
Client generation fails with `object has no key "UserBase"`

### ISSUE THREAD

**Issue by andrewkavalionak** [go-swagger/go-swagger#1065]
## Problem statement
Client generation fails with `object has no key "UserBase"`
## Swagger specification
2.0
## Steps to reproduce
1. Create two files
first.yml
```
swagger: '2.0'

info:
  version: '2.0'
  title: test
  description: test
  license:
    name: N/A

basePath: /v2

schemes:
  - http
  - https

consumes:
  - application/json
produces:
  - application/json

paths:
  /user:
    get:
      operationId: getUser
      description: test
      responses:
        200:
          description: succes
          schema:
            $ref: 'second.yml#/definitions/User'

```
/second.yml
```
swagger: '2.0'

info:
  version: '2.0'
  title: test
  description: test
  license:
    name: N/A

basePath: /v2

schemes:
  - http
  - https

consumes:
  - application/json
produces:
  - application/json

definitions:
  UserBase:
    type: object
    properties:
      id:
        type: integer
        format: uint64

  User:
    allOf:
    - $ref: '#/definitions/UserBase'
    - properties:
        createdAt:
          type: string
          format: date-time
        updatedAt:
          type: string
          format: date-time
```
2. run swagger generate client -f /Users/andrew/Workspace/go/src/github.com/go-swagger/example/first.yml -t /Users/andrew/Workspace/go/src/github.com/go-swagger/example/gen --skip-validation
3. swagger returns `object has no key "UserBase"`
## Environment
swagger version: latest from the master `220a68f31437616b53c032a5f82384450f82cf4e`
go version: go version go1.8.3 darwin/amd64
OS:   MacOS Sierra 10.12.5

**Comment by casualjim** [comment]
can you use `swagger flatten <spec>` before generating and see if the problem persists?

**Comment by andrewkavalionak** [comment]
yes, i tried this as well and for this case i was able to generate client from flattened spec, but if you add a reference to the properties it fails again with similar error
```
swagger: '2.0'

info:
  version: '2.0'
  title: test
  description: test
  license:
    name: N/A

basePath: /v2

schemes:
  - http
  - https

consumes:
  - application/json
produces:
  - application/json

definitions:
  UserBase:
    type: object
    properties:
      id:
        type: integer
        format: uint64

  ContactEmail:
    type: object
    properties:
      email:
        type: string

  User:
    allOf:
    - $ref: '#/definitions/UserBase'
    - properties:
    - properties:
        contactEmails:
          type: array
          items:
            $ref: '#/definitions/ContactEmail'
        createdAt:
          type: string
          format: date-time
        updatedAt:
          type: string
          format: date-time
```
small change to the second file. Just added 
```
contactEmails:
          type: array
          items:
            $ref: '#/definitions/ContactEmail'
```

and the error is
`schema analysis [#/definitions/user/allOf/2/properties/contactEmails]: object has no key "ContactEmail"`

Do you have any ideas?

**Comment by casualjim** [comment]
after flatten you can try to expand the spec `swagger expand`
then you should have a fully expanded spec for which you can generate code

**Comment by andrewkavalionak** [comment]
for the second case i'm unable to flatten 
`swagger flatten /Users/andrew/Workspace/go/src/github.com/go-swagger/example/first.yml -o /Users/andrew/Workspace/go/src/github.com/go-swagger/example/flatten-spec.json`
it fails with the error `schema analysis [#/definitions/user/allOf/1/properties/contactEmails]: object has no key "ContactEmail"`

I tried to expand and it worked.

**Comment by andrewkavalionak** [comment]
however if i modify specs a bit more `expand` stops working
**first.yml**
```
swagger: '2.0'

info:
  version: '2.0'
  title: test
  description: test
  license:
    name: N/A

basePath: /v2

schemes:
  - http
  - https

consumes:
  - application/json
produces:
  - application/json

paths:
  /user:
    get:
      operationId: getUser
      description: test
      responses:
        200:
          $ref: 'second.yml#/responses/userResponse'
```
**second.yml**
```
swagger: '2.0'

info:
  version: '2.0'
  title: test
  description: test
  license:
    name: N/A

basePath: /v2

schemes:
  - http
  - https

consumes:
  - application/json
produces:
  - application/json

responses:
  userResponse:
    description: succes
    schema:
      $ref:  '#/definitions/User'

definitions:
  UserBase:
    type: object
    properties:
      id:
        type: integer
        format: uint64

  ContactEmail:
    type: object
    properties:
      email:
        type: string

  User:
    allOf:
    - $ref: '#/definitions/UserBase'
    - properties:
        contactEmails:
          type: array
          items:
            $ref: '#/definitions/ContactEmail'
        createdAt:
          type: string
          format: date-time
        updatedAt:
          type: string
          format: date-time
```
swagger expand fails with `object has no key "User"` error

can you give me an idea where should i look at so I can maybe try to help fix that?

**Comment by casualjim** [comment]
expander is here:
https://github.com/go-openapi/spec/blob/master/expander.go

flatten is here:
https://github.com/go-openapi/analysis/blob/master/flatten.go

you can run processes with DEBUG=1 and SWAGGER_DEBUG=1 that will get you a lot more output

**Comment by EleanorRigby** [comment]
looks similar to issue # 890. 
The fix for 890 is on the way. I will test and see if it resolves this one too.

**Comment by reidrac** [comment]
I'm also affected by this; my case is the same @andrewkavalionak exposed: one YAML file with no definitions importing all the definitions from a second YAML file.

I'm implementing a dynamic API server and it loads the spec just fine (it opens and loads the definitions from the second file).

Is the client using a different mechanisms to load the spec?

Also: is flatten a super-set of expand?

I run expand + flatten and generated the client code, but I'm not sure the generated models are correct (28 models when I believe I only defined 3).

**Comment by EleanorRigby** [comment]
@reidrac : can u share your swagger spec (The two files you mentioned)? or a snippet which helps me to reproduce the case?

Flatten basically makes all your  references local.
Expand replaces the references with the intended content. IIRC, the expander works on local references. like `#/definitions`

**Comment by reidrac** [comment]
@EleanorRigby these are my current spec. Is a test app to evaluate go-swagger.

This is swagger.yaml:
```
swagger: '2.0'
info:
  version: 0.1.0
  title: Poll API
schemes:
  - http
consumes:
  - "application/json"
produces:
  - "application/json"
paths:
  /:
    get:
      tags: ["polls"]
      operationId: getAllPolls
      parameters:
        - name: limit
          in: query
          description: Limit the response to the last n polls.
          type: integer
          format: int32
          required: false
      responses:
        '200':
          description: OK
          schema:
            type: array
            items:
              $ref: "models.yaml#/definitions/poll"
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
    post:
      tags: ["polls"]
      operationId: addPoll
      parameters:
        - name: poll
          in: body
          schema:
            $ref: "models.yaml#/definitions/poll"
      responses:
        '201':
          description: Created
          schema:
            $ref: "models.yaml#/definitions/poll"
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
  /{id}:
    parameters:
      - type: integer
        format: int32
        name: id
        in: path
        required: true
    get:
      tags: ["polls"]
      operationId: getPoll
      responses:
        '200':
          description: OK
          schema:
            $ref: "models.yaml#/definitions/poll"
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
    put:
      tags: ["polls"]
      operationId: updatePoll
      parameters:
        - name: poll
          in: body
          schema:
            $ref: "models.yaml#/definitions/poll"
      responses:
        '200':
          description: OK
          schema:
            $ref: "models.yaml#/definitions/poll"
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
    delete:
      tags: ["polls"]
      operationId: deletePoll
      responses:
        '204':
          description: Deleted
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
  /{id}/{choice_id}:
    parameters:
      - type: integer
        format: int32
        name: id
        in: path
        required: true
      - type: integer
        format: int32
        name: choice_id
        in: path
        required: true
    put:
      tags: ["polls"]
      operationId: voteChoice
      responses:
        '200':
          description: OK
          schema:
            $ref: "models.yaml#/definitions/poll"
        default:
          description: error
          schema:
            $ref: "models.yaml#/definitions/error"
```

This is models.yaml:
```
swagger: '2.0'
definitions:
  poll:
    type: object
    required:
      - question
      - choices
    properties:
      id:
        type: integer
        format: int32
        readOnly: true
      question:
        type: string
        minLength: 1
        maxLength: 200
      pubDate:
        type: string
        format: date-time
        readOnly: true
      choices:
        type: array
        items:
          $ref: '#/definitions/choice'
  choice:
    type: object
    required:
      - choiceText
    properties:
      id:
        type: integer
        format: int32
        readOnly: true
      choiceText:
        type: string
        minLength: 1
        maxLength: 200
      votes:
        type: integer
        format: int32
        readOnly: true
  error:
    type: object
    required:
      - message
    properties:
      code:
        type: integer
        format: int64
      message:
        type: string
```

I run:
```
$ swagger generate client -f swagger.yaml -t ./restclient
2017/10/04 09:20:17 building a plan for generation
2017/10/04 09:20:17 planning definitions
2017/10/04 09:20:17 planning operations
object has no key "choice"
```

And that's all I know at this point.

Thanks for your help.

**Comment by reidrac** [comment]
Looks like is the same problem you have when run flatten:
```
$ swagger flatten swagger.yaml
2017/10/04 14:44:17 schema analysis [#/definitions/poll/properties/choices]: object has no key "choice"
```

(I've been reading the client generation and it does expand/flatten)

**Comment by EleanorRigby** [comment]
@reidrac : I will look into it. Will try to fix or share findings by eow.
Thanks for sharing.

**Comment by reidrac** [comment]
Looking at the schema I posted, if I inline the Choice schema instead of using a reference in the choices array of Poll; flatten works OK (the resulting JSON looks fine, although "x-go-gen-location": "models" doesn't seem to be consistent.

When I try to generate the client with these YAML files, Swagger doesn't report any error, but the models aren't generated either.

When I use --dump-data flag, I can see "Models" is null, although "IncludeModel" is true. Perhaps that should trigger an error.

**Comment by reidrac** [comment]
Generate client doesn't resolve external references as it is; my previous comment re: expand was red herring (expand is called when the Spec is validated prior to generation).

So there's no way it'll work without running flatten first (and then there's an error that I avoided by inlining Choice definition), and that requires using an intermediate JSON file. Summing up: you can't generate a client with a YAML file that has external refs in the definitions.

@EleanorRigby Can you confirm I'm in the right path? Thanks!

**Comment by EleanorRigby** [comment]
@reidrac : looking into it now.

**Comment by EleanorRigby** [comment]
@reidrac : Initial analysis points to same conclusion that flatten needs to be run first which makes all schema refs local. Also, the choice has to be inline (as the way current code works) Does that unblock you to work?
The thing that confuses me is that flatten is run while generating server and client within the code unless one specifies --skip-flatten flag. The flattening should have happened. I am continuing my analysis further.

**Comment by EleanorRigby** [comment]
@reidrac : After some more digging, I now see what is happening inside the code. The current binary does not resolve any local reference inside the external reference (external ref --> Poll has a local ref --> Choice under Choices array)

The solution to this problem seems non-trivial and might take some thinking and time to fix. The best work around I can suggest currently is to inline the Choice (which you did) and flatten the spec.

The expansion of spec before generation also works perfect. The number of models generated are huge because references are used in multiple operations(That is how swagger tool behaves currently). Ideally , the `AddPollParamsBodyChoicesItems` and `UpdatePollParamsBodyChoicesItems` are the same structures but as far as I know the swagger binary does this so as to accomodate changes caused by complex definitions involving allof , anyof etc..

It would be helpful if you can think/share  a solution or design to handle these recursive references distributed within and in external files.

**Comment by EleanorRigby** [comment]
Adding on to the discussion for future work (proposed idea)

- In flatten function we need a DAG of all possible references
      a. Starting the DAG creation from root doc specified.
      b. All the local refs must carry a file string to indicate the local file name where these local refs have expanded definitions
- The references will be resolved in a topological sort manner
-  All the references which get free of other ref dependencies will have a map of refname --> absolute resolved definition values.
- The final values will be substituted in root document.

**Comment by reidrac** [comment]
@EleanorRigby thanks for you help!

* `--skip-flatten` flag is not available in swagger 0.12.0 (?), which version should I be using? 
* this is a PoC for a larger project and we were hopping to work with the YAML file, using shared definitions in different API endpoints (that's why we are exploring refs to external files); currently our working solution uses the mixin approach, that means we work with YAML but in practice we consume the "mixed" JSON file. The YAML files are OK to share with 3rd party to generate clients in other languages.
* we're using the dynamic API approach at the server; so no code generation there. I'm confused because the YAML file I posted here seems to work on the server just fine (?). I may be missing something here or the definitions are processed differently.

Unfortunately I can't suggest a solution. The time I've spent reading go-swagger code I found it a bit complicated; I miss more consistent logging, currently is quite hard to follow what's going on if you're not familiar with the codebase.

At a very high level, flatten + expand sounds like a reasonable strategy, but flatten should detect when a definition includes a reference to another definition; perhaps running expand on any spec referenced by the main Spec could do it (so Choice would be inlined automatically, right?).
